plat/kvm/arm/smp.c

```diff
@@ -0,0 +1,401 @@
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <libfdt.h>
+#include <kvm/config.h>
+#include <kvm-arm/smp.h>
+#include <kvm/intctrl.h>
+#include <arm/cpu.h>
+#include <kvm-arm/smp_defs.h>
+#include <uk/plat/io.h>
+#include <uk/plat/lcpu.h>
+#include <arm/psci.h>
+#include <ofw/fdt.h>
+#include <uk/arch/spinlock.h>
+#include <uk/plat/irq.h>
+#include <arm/time.h>
+#include <uk/plat/time.h>
+
+#define FDT_SIZE_CELLS_DEFAULT 0
+#define FDT_ADDR_CELLS_DEFAULT 2
+
+struct arm64_cpu cpus[CONFIG_UKPLAT_LCPU_MAXCOUNT];
+
+static __u32 bsp_cpu_id;
+static unsigned int smp_numcores;
+void _lcpu_start(struct arm64_cpu *cpu);
+void __noreturn _lcpu_entry_default(void);
+
+static __lcpuid ukplat_lcpu_idx(void);
+
+static int arm64_lcpu_ipi_run_handler(void *args __unused, struct __regs *regs)
+{
+	struct ukplat_lcpu_func *i;
+	struct uk_list_head *fnlist;
+
+	__u32 this_cpu_idx = ukplat_lcpu_idx();
+
+	cpus[ukplat_lcpu_id()].state = LCPU_STATE_BUSY;
+	ukarch_spin_unlock(&cpus[this_cpu_idx].lock);
+
+	fnlist = &cpus[this_cpu_idx].fnlist;
+
+	uk_list_for_each_entry(i, fnlist, lentry) {
+		i->fn(regs, i);
+		if (i == uk_list_first_entry(fnlist,
+				struct ukplat_lcpu_func, lentry))
+			break;
+	}
+
+	intctrl_ack_irq(UKPLAT_ARM64_IPI_RUN_IRQ);
+
+	return 1;
+}
+
+int smp_init(void)
+{
+	int fdt_cpu;
+	const fdt32_t *naddr_prop, *size_prop, *id_reg;
+	const struct fdt_property *cpu_prop;
+	int naddr, size = 0, cell;
+	unsigned int cpu_idx;
+	int subnode;
+
+	/* Init the info for the bootstraping cpu */
+	bsp_cpu_id = ukplat_lcpu_id();
+	cpus[0].id = bsp_cpu_id;
+	cpus[0].state = LCPU_STATE_BUSY;
+	ukarch_spin_init(&cpus[0].lock);
+	uk_pr_info("Bootstrapping processor has the ID %d\n", bsp_cpu_id);
+
+	/* Init the secondary cpus vector */
+	for (cpu_idx = 1; cpu_idx < CONFIG_UKPLAT_LCPU_MAXCOUNT; cpu_idx++) {
+		cpus[cpu_idx].id = (__u32) -1;
+		cpus[cpu_idx].state = LCPU_STATE_OFFLINE;
+	}
+	smp_numcores = 0;
+
+	/* Search for cpus in DTB */
+	fdt_cpu = fdt_path_offset(_libkvmplat_cfg.dtb, "/cpus");
+	if (fdt_cpu < 0) {
+		uk_pr_err("cpus node is not found in device tree\n");
+		return -1;
+	}
+
+	/* Get address cells */
+	naddr_prop = fdt_getprop(_libkvmplat_cfg.dtb,
+					fdt_cpu, "#address-cells", &cell);
+	if (naddr_prop == NULL) {
+		naddr = FDT_ADDR_CELLS_DEFAULT;
+		uk_pr_warn("Using default value 2 for cpu address-cells.\n");
+	} else {
+		naddr = fdt32_to_cpu(naddr_prop[0]);
+	}
+	if (naddr < 0 || naddr >= FDT_MAX_NCELLS) {
+		uk_pr_err("Invalid adress-cells size!\n");
+		return -1;
+	}
+
+	/* Get size cells */
+	size_prop = fdt_getprop(_libkvmplat_cfg.dtb, fdt_cpu,
+			"#size-cells", &cell);
+	if (size_prop == NULL) {
+		naddr = FDT_SIZE_CELLS_DEFAULT;
+		uk_pr_warn("Using default value of 0 for cpu size-cells.\n");
```

Comment:
0 is the default value when no `size-cells` property is given. We shouldn't throw an error, since according to the docs, 0 is the `size-cells` value for cpu nodes:
```
Value shall be 0. Specifies that no size is required in the reg property in children of this node.
```
I can't see any benefit in the user setting this property explicitly, so I think we could safely remove this message.
However, I am thinking right now that maybe an error message should be thrown if the device tree specifies a `size-cells` property greater than 0 for cpus nodes. What do you think?

Reply:
Yup. Let's get rid of it then.